_studio/shared/src/mfx_umc_mjpeg_vpp.cpp

```diff
@@ -171,75 +124,51 @@ UMC::Status mfx_UMC_FrameAllocator_D3D_Converter::InitMfx(UMC::FrameAllocatorPar
 
 mfxStatus mfx_UMC_FrameAllocator_D3D_Converter::StartPreparingToOutput(mfxFrameSurface1 *surface_work,
                                                                        UMC::FrameData* in,
-                                                                       const mfxVideoParam * par,
-                                                                       VideoVppJpegD3D9 **pCc,
+                                                                       const mfxVideoParam *par,
+                                                                       VideoVppJpegD3D9 *pCc,
                                                                        mfxU16 *taskId,
                                                                        bool isOpaq)
 {
     UMC::AutomaticUMCMutex guard(m_guard);
 
     mfxStatus sts = MFX_ERR_NONE;
-    bool isD3DToSys = false;
+
+    mfxHDLPair pPair;
+    if(isOpaq)
+        sts = m_pCore->GetFrameHDL(surface_work->Data.MemId, (mfxHDL*)&pPair);
+    else
+        sts = m_pCore->GetExternalFrameHDL(surface_work->Data.MemId, (mfxHDL*)&pPair);
 
     if(par->mfx.FrameInfo.PicStruct == MFX_PICSTRUCT_PROGRESSIVE)
```

Comment:
From refactoring point of view it makes sense to extract this surface processing (obtaining memId, adjusting width/height etc.) into separate method and then call it from both progressive and interlace branches.

Reply:
I unable to unify adjusting width/height.

Reply:
```
inline void AdjustSize(mfxFrameInfo& info, int rotation)
{
    if (!(rotation == MFX_ROTATION_0 || rotation == MFX_ROTATION_180))
        std::swap(info.CropW, info.CropH);

    if (info.PicStruct != MFX_PICSTRUCT_PROGRESSIVE)
        info.CropH /= 2;
}
///...
mfxStatus mfx_UMC_FrameAllocator_D3D_Converter::StartPreparingToOutput(mfxFrameSurface1 *surface_work, ...)
{
    ///...
    for (int i = 0; i < surface_work->Info.PicStruct != MFX_PICSTRUCT_PROGRESSIVE; ++i)
    {
        //assuming `m_frameDataInternal.GetSurface` copies `FrameInfo`
        mfxFrameSurface1 surface = m_frameDataInternal.GetSurface(in[i]->GetFrameMID()); 
        AdjustSize(surface.Info, par->mfx.Rotation);
    }
}

```

Reply:
Thanks, got the loop & array idea. I'm not sure the adjusting deserve a special function. It seems `BeginHwJpegProcessing` can be unified in this way as well.